libraries/AP_Avoidance/AP_Avoidance.cpp

```diff
@@ -0,0 +1,598 @@
+    // @User: Advanced
+    AP_GROUPINFO("W_DIST_XY",     8, AP_Avoidance, _warn_distance_xy,    300),
+
+    // @Param: DIST_F_XY
+    // @DisplayName: Distance Fail XY
+    // @Description: Closest allowed projected distance before BEHAVIOUR_F is undertaken
+    // @User: Advanced
+    AP_GROUPINFO("F_DIST_XY",     9, AP_Avoidance, _fail_distance_xy,    100),
+
+    AP_GROUPEND
+};
+
+const char *AP_Avoidance::_state_names[] = { "CLEAR", "WARN", "FAIL" };
+
+AP_Avoidance::AP_Avoidance(AP_AHRS &ahrs, AP_ADSB &adsb) :
+    _ahrs(ahrs),
+    _adsb(adsb)
+{
+    AP_Param::setup_object_defaults(this, var_info);
+    _instance = this;
+}
+
+/*
+ * Initialize variables and allocate memory for array
+ */
+void AP_Avoidance::init(void)
+{
+    if (_samples == NULL) {
+        _samples = new AP_Avoidance::Sample[_samples_max];
+
+        if (_samples == NULL) {
+            // dynamic RAM allocation of _samples[] failed, disable gracefully
+            hal.console->printf("Unable to initialize Avoidance sample list\n");
+            // disable ourselves to avoid repeated allocation attempts
+            _enabled.set(0);
+            return;
+        }
+    }
+    _sample_count = 0;
+    _last_state_change_ms = 0;
+    _old_state = STATE_CLEAR;
+    _gcs_cleared_messages_first_sent = std::numeric_limits<uint32_t>::max();
+    _current_most_serious_threat = -1;
+}
+
+/*
+ * de-initialize and free up some memory
+ */
+void AP_Avoidance::deinit(void)
+{
+    if (_samples != nullptr) {
+        delete [] _samples;
+        _samples = nullptr;
+    }
+    _sample_count = 0;
+}
+
+bool AP_Avoidance::check_startup()
+{
+    if (!_enabled) {
+        if (_samples != nullptr) {
+            deinit();
+        }
+        // nothing to do
+        return false;
+    }
+    if (_samples == nullptr)  {
+        init();
+    }
+    return _samples != nullptr;
+}
+
+// vel is north/east/down!
+void AP_Avoidance::add_sample(const uint32_t sample_timestamp_ms,
+                              const MAV_COLLISION_SRC src,
+                              const uint32_t src_id,
+                              const Location &loc,
+                              const Vector3f &vel)
+{
+    if (! check_startup()) {
+        return;
+    }
+    uint32_t oldest_timestamp = std::numeric_limits<uint32_t>::max();
+    uint8_t oldest_index = 255; // avoid compiler warning with initialisation
+    int16_t index = -1;
+    uint8_t i;
+    for (i=0; i<_sample_count; i++) {
+        if (_samples[i].src_id == src_id &&
+            _samples[i].src == src) {
+            // pre-existing sample found; we will update its information
+            index = i;
+            break;
+        }
+        if (_samples[i].timestamp_ms < oldest_timestamp) {
+            oldest_timestamp = _samples[i].timestamp_ms;
+            oldest_index = i;
+        }
+    }
+    if (index == -1) {
+        // existing sample not found.  See if we can store it anyway:
+        if (i <_samples_max) {
```

Comment:
if _samples_max (AVOID_OBS_MAX) changes during runtime, which is very likely, then you could have a memory problem here.